// @rbx/core/dist/device/index.js
export var Platform;
(function(Platform) {
    Platform[Platform["Windows"] = 0] = "Windows";
    Platform[Platform["macOS"] = 1] = "macOS";
    Platform[Platform["Linux"] = 2] = "Linux";
    Platform[Platform["Unix"] = 3] = "Unix";
    Platform[Platform["iOS"] = 4] = "iOS";
    Platform[Platform["Android"] = 5] = "Android";
    Platform[Platform["Unknown"] = 6] = "Unknown";
})(Platform || (Platform = {}));
// credit: https://medium.com/@justenphelps/mobile-os-detection-1f14b46565cc
export function getCurrentPlatform() {
    if (typeof navigator !== 'undefined') {
        var platform = navigator.platform;
        if (platform.includes('Win')) {
            return Platform.Windows;
        }
        if (platform.includes('Mac')) {
            return Platform.macOS;
        }
        if (platform.includes('Linux')) {
            return Platform.Linux;
        }
        if (platform.includes('X11')) {
            return Platform.Unix;
        }
        if (platform.includes('iPhone')) {
            return Platform.iOS;
        }
        if (platform.includes('Android')) {
            return Platform.Android;
        }
        return Platform.Unknown;
    }
    // on server side
    return null;
}
export default {
    getCurrentPlatform: getCurrentPlatform
};
