// @rbx/core/dist/cursorPagination/PaginationCache.js
var PaginationCache = /** @class */ (function() {
    function PaginationCache(pageSize) {
        this.pageSize = pageSize;
        this.cache = {};
    }
    PaginationCache.prototype.getPage = function(cacheKey, pageNumber) {
        var cacheList = this.cache[cacheKey];
        if (cacheList) {
            return cacheList.slice((pageNumber - 1) * this.pageSize, pageNumber * this.pageSize);
        }
        return [];
    };
    PaginationCache.prototype.getLength = function(cacheKey) {
        var cacheList = this.cache[cacheKey];
        if (cacheList) {
            return cacheList.length;
        }
        return 0;
    };
    PaginationCache.prototype.append = function(cacheKey, values) {
        if (!this.cache[cacheKey]) {
            this.cache[cacheKey] = [];
        }
        this.cache[cacheKey] = this.cache[cacheKey].concat(values);
    };
    PaginationCache.prototype.removeAfterIndex = function(cacheKey, index) {
        if (this.cache[cacheKey]) {
            this.cache[cacheKey] = this.cache[cacheKey].slice(0, index);
        }
    };
    PaginationCache.prototype.removeAtIndex = function(cacheKey, pageNumber, index) {
        if (this.cache[cacheKey]) {
            this.cache[cacheKey].splice((pageNumber - 1) * this.pageSize + index, 1);
        }
    };
    PaginationCache.prototype.clear = function(cacheKey) {
        delete this.cache[cacheKey];
    };
    PaginationCache.prototype.setPageSize = function(pageSize) {
        this.pageSize = pageSize;
    };
    return PaginationCache;
}());
export default PaginationCache;
